### ISSUE TITLE
Error when launching through angular

### ISSUE THREAD

**Issue by roddone** [isomorphic-git/isomorphic-git#597]
Hi, 

I'm using isomorphic-git in browser through an Angular 7 project. I successfully installed the project with `npm install --save isomorphic-git` but when launching my project with `ng serve` I encouter the following error : 

`ERROR in node_modules/isomorphic-git/src/index.d.ts(13,11): error TS2580: Cannot find name 'Buffer'. Do you need to install type definitions for node? Try `npm i @types/node`.
node_modules/isomorphic-git/src/index.d.ts(115,13): error TS2580: Cannot find name 'Buffer'. Do you need to install type definitions for node? Try `npm i @types/node`.node_modules/isomorphic-git/src/index.d.ts(561,11): error TS2580: Cannot find name 'Buffer'. Do you need to install type definitions for node? Try `npm i @types/node`.`

`@types/node` is already installed and I tried updating it does not solve my problem.

The error is thrown when I import isomorphic-git in my components : `import { clone } from 'isomorphic-git'`

Is it a known problem and do you know if the library is compatible with angular projects ?

Regards, 

Romain.

**Comment by wmhilton** [comment]
Darn. I really have no idea... I'm just googling "Typescript TS2580" but I found this: https://github.com/aws/aws-sdk-js/issues/1271

Maybe that'll help? Apparently there's a setting in a file called "`tsconfig.app.json`" that needs to be turned on for it to see `@types/node` in the client (browser) code.

**Comment by wmhilton** [comment]
Good news it seems to work with Angular 7 in this online sandbox: https://stackblitz.com/edit/angular-pdbaet

**Comment by roddone** [comment]
It works in the sandbox but I tried to download the project and ran it into my local dev environment and I got the same errors as above. I assume the sandbox has a different build process.

It seems isomorphic-git package requires node libraries that are note compatibles with angular because they are "Node" packages : https://github.com/angular/angular-cli/issues/11130
My team and I tried several workarounds but none of them worked because everytime isomorphic package is loaded, it references one of these "node" packages and it kills angular build process.

**Comment by wmhilton** [comment]
> It works in the sandbox but I tried to download the project and ran it into my local dev environment and I got the same errors as above. I assume the sandbox has a different build process.

Hmm you're right! That is frustrating.

This appears to be a popular frustration in the Angular community: https://github.com/angular/angular-cli/issues/9827

I've tried adding Buffer to `src/polyfills.ts` but that didn't seem to work either.

**Comment by wmhilton** [comment]
OK... I have it working! You lose all the type checking and don't get any tree-shaking of the bundle but if you do:

```ts
import * as git from 'isomorphic-git/dist/bundle.umd.min.js';
```

it works 😀

**Comment by roddone** [comment]
> For the record, if you update `tsconfig.app.json` with `types: ["node"]` then you _do_ get passed the Buffer error ... and if you delete the `corsProxy` argument (because apparently I forgot to add that to the TS definitions for getRemoteInfo)... then you finally get to some _meaningful_ errors:
> 
> ```
> ERROR in ./node_modules/simple-get/index.js
> Module not found: Error: Can't resolve 'http' in 'C:\Downloads\angular-pdbaet\node_modules\simple-get'
> ERROR in ./node_modules/simple-get/index.js
> Module not found: Error: Can't resolve 'https' in 'C:\Downloads\angular-pdbaet\node_modules\simple-get'
> ERROR in ./node_modules/globalyzer/src/index.js
> Module not found: Error: Can't resolve 'os' in 'C:\Downloads\angular-pdbaet\node_modules\globalyzer\src'
> ERROR in ./node_modules/globalyzer/src/index.js
> Module not found: Error: Can't resolve 'path' in 'C:\Downloads\angular-pdbaet\node_modules\globalyzer\src'
> ERROR in ./node_modules/isomorphic-git/dist/for-future/isomorphic-git/index.js
> Module not found: Error: Can't resolve 'path' in 'C:\Downloads\angular-pdbaet\node_modules\isomorphic-git\dist\for-future\isomorphic-git'
> ERROR in ./node_modules/isomorphic-git/dist/for-future/isomorphic-git/index.js
> Module not found: Error: Can't resolve 'stream' in 'C:\Downloads\angular-pdbaet\node_modules\isomorphic-git\dist\for-future\isomorphic-git'
> ```
> To me this indicates that the library is relying on five node builtins that are normally shimmed by `webpack`: 'http', 'https', 'os', 'path', and 'stream'.
> 
> The 'http' and 'https' dependencies will likely go away when I abstract the `httpRemoteHelper` into its own plugin. 'os' is being used by 'globalyzer'... not sure what for, but it might need to be shimmed. 'path' is also used by globalyzer but is easy to shim. And 'stream'... I shouldn't be relying on stream, I should be using `readable-stream` like every good node citizen, I'll have to check what's up with that.
> 
> So that's the plan then. Phase out `simple-get`, replace 'stream' with 'readable-stream', and probably talk to the author of `globalyzer` and see if we can make it play ball with Angular 7.

Most of my attempts to solve the problem ended like this 😂 Thank you for this fine analysis !

**Comment by wmhilton** [comment]
@roddone How about the "solution" of using the pre-built bundle like in https://github.com/isomorphic-git/isomorphic-git/issues/597#issuecomment-443271254 ? Does that work for your project?

Meanwhile, I've knocked out the 'stream' dependency! So we're 20% closer. Just 'path', 'http', 'https', and 'os'.

**Comment by roddone** [comment]
@wmhilton yes we are doing this for now and it seems to work. Since it's just the build process that fails, using the library with the pre-built bundle without typings works like a charm !

**Comment by wmhilton** [comment]
### Status update:

We're much closer to working out-of-the-box with Angular 7! Which is also means were closer to achieving #698.

I was able to get our [demo app](https://stackblitz.com/edit/angular-1da8wb) running locally with only the following modifications:

`package.json`
```
...
  "dependencies": {
...
    "buffer": "^5.2.1",
    "path-browserify": "^1.0.0"
  },
...
```

`polyfills.ts`
```ts
...
(window as any).global = window;
(window as any).process = {};
import { Buffer } from 'buffer';
(window as any).Buffer = Buffer;
```

`tsconfig.json`
```
...
    "lib": [
      "esnext.asynciterable",
...
    ]
...
```

`src/tsconfig.app.json`
```
...
  "compilerOptions": {
...
    "types": ["node"],
    "paths": {
      "path": ["node_modules/path-browserify"]
    }
  }
...
```

**Comment by alemarre** [comment]
Hi,

Thx wmhilton, it's working with those small modifications and the bundled library.

If this is useful to anyone, you can still add some typings using Typescript keyword `ReturnType` : 

```ts
import * as git from 'isomorphic-git/dist/bundle.umd.min.js';
// Won't be compiled since only typings are used from this import
import * as GitTypes from 'isomorphic-git';

[...]
  showLogs(): ReturnType<typeof GitTypes.log> {
    return this.initRepo
      .then(() => git.log({ dir: '/', depth: 5, ref: 'master' }));
  }

```

**Comment by OrangeDrangon** [comment]
How much of this is still relevant? Are the changes above still required or has this progressed!

**Comment by tgirotto** [comment]
tested on Angular 8 by:
1) importing "node_modules/isomorphic-git/dist/bundle.umd.min.js" in angular.json,
2) declaring 

```
function startGit() {
      $(document).ready(function() {
           BrowserFS.configure({ fs: "IndexedDB", options: {} }, function (err) {
           if (err) return console.log(err);
               window.fs = BrowserFS.BFSRequire("fs");
               git.plugins.set('fs', window.fs);
           });
     })
}
```
in src/assets/somegitfile.js

3) importing into any controller the following and then call it from anywhere in that controller.

```
declare function startGit():any`
```

**Comment by wmhilton** [comment]
Good to know the minified version still works. Hopefully one day the tree-shakable version will work as well. I still haven't eliminated `Buffer` from the code yet though.

**Comment by DZiso** [comment]
for angular 8 i had to replace my polyfill.ts with:

`(window as any).global.Buffer = (window as any).global.Buffer || require('buffer').Buffer;
`

instead of this:
```
import { Buffer } from 'buffer';
(window as any).Buffer = Buffer;
```

**Comment by crutchcorn** [comment]
Another place where this is important: React Native using Typescript.

I'm hoping to build an open-source mobile Git client and want to use isomorphic git but am hitting this same issue (and am working around it by using the UMD release)

**Comment by MikeJerred** [comment]
Only these changes are required with Angular 13:

`package.json`

```
"dependencies": {
  "buffer": "^6.0.3"
},
...
```

`polyfills.ts`

```ts
import { Buffer } from 'buffer';
(window as any).Buffer = Buffer;
```

**Comment by jcubic** [comment]
@MikeJerred thanks, there're at least two issues about buffer and webpack. This needs to be added to FAQ. With all working solutions.
